docs/src/keymap-upgrade.js

```diff
@@ -0,0 +1,55 @@
+import { Codes, Behaviors } from "./data/keymap-upgrade";
+
+// The start of a behvioar like "&kp" isn't a word boundary,
+// so also match following the start of a line or whitespace.
+const WordBegin = "(?:\\b|^|(?<=\\s))";
+const WordEnd = "\\b";
+const UnimportantWhitespace = "( +)(?!\\S)";
+
+export function upgradeKeymap(text) {
+  text = upgradeBehaviors(text);
+  text = upgradeKeycodes(text);
+  return text;
+}
+
+function upgradeBehaviors(text) {
+  for (const [deprecated, replacement] of Object.entries(Behaviors)) {
+    text = replaceWord(text, deprecated, replacement);
+  }
+
+  return text;
+}
+
+function upgradeKeycodes(text) {
+  for (const [deprecated, replacement] of Object.entries(Codes)) {
+    if (replacement) {
+      text = replaceWord(text, deprecated, replacement);
+    } else {
+      text = replaceWord(
+        text,
+        `&kp\\s+${deprecated}`,
+        `&none /* Cannot upgrade ${deprecated} */`
+      );
+    }
+  }
+
+  return text;
+}
+
+function replaceWord(text, word, replacement) {
+  // Match the word and all unimportant whitespace following it
+  const pattern = new RegExp(
+    `${WordBegin}(${word})${WordEnd}(${UnimportantWhitespace})?`,
+    "gm"
+  );
+  return text.replace(pattern, (match, word, spaceAfter) => {
+    return getAlignedReplacement(word, replacement, spaceAfter?.length ?? 0);
+  });
+}
```

Comment:
I noticed it screwed up some of the ascii layouts that people use.  We could probably minimize that by adding rudimentary context wrapping to the RegExp.  Something along the lines of:
```
`(?:bindings\\s?=\\s?<[^>]+)${WordBegin}(${word})${WordEnd}(${UnimportantWhitespace})?(?:[^>]+>)`
```
It should handle `sensor-bindings` also.  What do you think please?

Reply:
I'd considered doing something like that, but if someone had this in their keymap, it'd break.
```
bindings = <
    /* < = > (whoops, the regex stops here)  */
    &kp LT &kp EQL &kp GT
>;
```
You'd need a smarter parser that's able to recognize comments, and while you can use regexes to build such a parser, regexes aren't suited for that on their own.

Reply:
That's true, however:
- most (all?) shields have their ascii comments *above* the `bindings`, not inside them
- comments inside `bindings` containing `>` are likely to be rare
- it's better *not* to convert, than to convert unrelated parts of the file that the user doesn't anticipate or notice changing

Any thoughts?

Reply:
The other approach to improve context is to include the behavior tags into the regexp, but it's probably more involved due to the numerous behaviors.

Reply:
I see three options here:
1. Do a simple word search and replace, which may convert unrelated parts of the file.
2. Use some naïve regex parsing to try to limit searches to specific parts of the file, which may fail to upgrade things that were supposed to be upgraded.
3. Use a proper parser.

In option 1, I believe the only things it might unexpectedly convert would be comments. Anything else would be seen by the preprocessor as an instance of one of the macros we're trying to upgrade, and thus should be upgraded too.

I personally see option 2 as worse because instead of unexpectedly converting some extra things that shouldn't have any effect on the function of the keymap, it might unexpectedly _not_ convert some things that do affect the function of the keymap.

Option 3 would obviously be the most reliable, but it is a lot harder. I tried that first using https://tree-sitter.github.io/tree-sitter/ but couldn't figure out how to resolve some ambiguities between the syntax for child nodes and properties. I could try porting Zephyr's DT parser to JavaScript instead of writing a parser from scratch.

Reply:
Whoops. I forgot I was still logged into my work account. If it wasn't already obvious, the above is also me.

Reply:
Yeah, that was my impression from what you've said before.  But I tried it on many files and on balance I found that only stuff inside the bindings usually needed changing, and any comments that were also changed really threw me off.  I'd imagine the vast majority of users will be the same as most maps are rudimentary.

Reply:
For my own reference later, https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/dts/dtlib.py is Zephyr's parser. It is not trivial, but a port of that to JS would also be useful for making a VS Code (or similar IDE) extension for autocomplete in keymaps.